src/node/historical_queries_utils.cpp

```diff
@@ -3,11 +3,158 @@
 
 #include "ccf/historical_queries_utils.h"
 
+#include "ccf/crypto/cose_verifier.h"
 #include "ccf/rpc_context.h"
 #include "ccf/service/tables/service.h"
 #include "kv/kv_types.h"
 #include "node/identity.h"
 #include "node/tx_receipt_impl.h"
+#include "service/tables/previous_service_identity.h"
+
+namespace
+{
+  using Endorsements = std::vector<std::vector<uint8_t>>;
+  struct FetchResult
+  {
+    std::optional<Endorsements> endorsements{std::nullopt};
+    bool retry{false};
+  };
+
+  static std::vector<ccf::CoseEndorsement> cose_endorsements_cache = {};
+
+  bool is_self_endorsement(const ccf::CoseEndorsement& endorsement)
+  {
+    return !endorsement.previous_version.has_value();
+  }
+
+  void validate_fetched_endorsement(
+    const std::optional<ccf::CoseEndorsement>& endorsement)
+  {
+    if (!endorsement)
+    {
+      throw std::logic_error("Fetched COSE endorsement is invalid");
+    }
+
+    if (!is_self_endorsement(*endorsement))
+    {
+      const auto [from, till] = ccf::crypto::extract_cose_endorsement_validity(
+        endorsement->endorsement);
+      const auto from_txid = ccf::TxID::from_str(from);
+      const auto till_txid = ccf::TxID::from_str(till);
+      if (
+        !till_txid || !endorsement->endorsed_till ||
```

Comment:
Missing a !from_txid before using it next line. I think splurging on individual has_value() checks, with individualised exception messages for each one, may be a worthwhile investment here.